File: java/src/main/java/com/mysqltest/test/MysqlTest.java
package com.mysqltest.test;
import java.sql.Connection;  
import java.sql.DriverManager;  
import java.sql.PreparedStatement;  
import java.sql.SQLException;  
import java.sql.Statement;
import java.sql.Date;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.List;
import java.util.ArrayList;

public class MysqlTest {  
	
	static int  count=10000000;//总次数
	//一定要写rewriteBatchedStatements参数，Mysql批量插入才性能才理想
	static String mySqlUrl="jdbc:mysql://127.0.0.1:3306/test?rewriteBatchedStatements=true";
	static String mySqlUserName="root";  
	static String mySqlPassword="cloudera";  
	
	//batchsize
	// static int[] divPoint={count,10000,1000,100,10,1};
	
    public static void main(String[] args) {  
        String[] tableName={"test1","test2","test3"};
        String inputPath="/mnt/mysqlTest/data.txt";
        // --------------一次性读完表-----------------------
        // List<tableStruct> values;
        // try{
        //     values=getData(inputPath,10000000);
        // }catch(IOException e){
        //     System.out.println("inputPath error");
        //     values=null;
        // }
        // --------load 和分批insert效果比较-----
        // for (String tab:tableName){
        //     // insertPerline(values,tab,"ignore");
        //     insertBatch(values,tab,"ignore",10000);
        //     // insertBatch(values,tab,"replace",10000);
        //     loadAll(tab,"ignore");
        //     loadAll(tab,"replace");
        // }

        // --------insert不同大小分批的效果--------
        // insertBatch(values,tab,"ignore",)
        // for (String tab:tableName){
        //     for(int point:divPoint){
        //         insertBatch(values,tab,"ignore",point);
        //     }
        // }

        // -------java和python的比较-------------
        // insertBatch(values,tableName[2],"ignore",100000);
        // loadAll(tableName[2],"ignore");
        // loadAll(tableName[2],"replace");
        // -------分批load-------------
        // loadAll(tableName[2],"ignore");
        // -------java connection-------------

        // mysql批量插入记录,div为划分的个数，每次重新建立连接
        try{
            insertBatchReConn(inputPath,tableName[2],"ignore",100);
        }catch(IOException e){
            System.out.println("inputPath error");
        }
    }  

    
    // 封装行数据
    private static class tableStruct{                 //消息头
        public String  name;          
        public String  birth;   
        public int id;        
        public int height;        
        public int weight;       
    }; 

    // 读取文件
    public static List<tableStruct> getData(String inputPath,int count)throws IOException
    {
        List<tableStruct> values=new ArrayList<tableStruct>();
        FileReader reader = new FileReader(inputPath);
        BufferedReader br = new BufferedReader(reader);          
        for (int i=0;i<count;i++){
            tableStruct value=new tableStruct();
            String[] str=new String[5];
            str=br.readLine().split("\\,");
            value.name=str[0].substring(1,9);
            value.birth=str[1].substring(1,11);
            value.id=Integer.parseInt(str[2]);
            value.height=Integer.parseInt(str[3]);
            value.weight=Integer.parseInt(str[4]);      
            values.add(value);
        }
        return values;
    }
    /**
     * 创建连接
     * @return
     */
    public static Connection getConn(String flag){
    	long a=System.currentTimeMillis();
        try {        
        	if("mysql".equals(flag)){
                Class.forName("com.mysql.jdbc.Driver");        
                Connection conn =  DriverManager.getConnection(mySqlUrl, mySqlUserName, mySqlPassword);     
                conn.setAutoCommit(false);  
                return conn;
        	}else{
        		System.out.println();
        		throw new RuntimeException("flag参数不正确,flag="+flag);
        	}
        }catch (Exception ex) {  
            ex.printStackTrace();  
        }finally{  
        	long b=System.currentTimeMillis();  
            System.out.println("创建连接用时"+ (b-a)+" ms"); 
        }
        return null;
    }
    /**
     * 关闭连接
     * @return
     */
    public static void close(Connection conn){
    	 try {  
             if(conn!=null){
            	 conn.close();  
             }
         } catch (SQLException e) {  
             e.printStackTrace();  
         }
    }
    /**
     * 删除旧数据
     * @return
     */
    public static void clear(Connection conn, String tableName){
    	try{
            Statement st=conn.createStatement();
            boolean bl=st.execute("truncate table "+tableName);
            conn.commit();
            st.close();
            System.out.println("执行清理操作："+(bl==false?"成功":"失败"));
    	}catch(Exception e){
    		e.printStackTrace();
    	}
    }
    /**
     * 打印信息
     * @return
     */
    public static void print(String key,long startTime,long endTime,int batchsize){
    	// System.out.println("每执行"+count+"次sql提交一次事务");
        double speed=batchsize*1000.0/(endTime-startTime);
    	System.out.println(key+",用时"+ (endTime-startTime)/1000.0+" s,平均每秒执行"+speed+"条");
    	System.out.println("----------------------------------");
    }
    /** 
     * mysql非批量插入记录 
     */  
    public static void insertPerline(List<tableStruct> values,String tableName,String mode){  
        String sql="insert "+mode+" into "+ tableName+ " values(?,?,?,?,?)";
        Connection conn=getConn("mysql");  
        clear(conn,tableName);
        try {        
            PreparedStatement prest = conn.prepareStatement(sql);        
            long a=System.currentTimeMillis();  
            for(int i=0;i<count;i++){
                prest.setString(1,values.get(i).name);
                prest.setString(2,values.get(i).birth);        
                prest.setInt(3,values.get(i).id);
                prest.setInt(4,values.get(i).weight);
                prest.setInt(5,values.get(i).height);        
                prest.execute();  
                // if(x%point==0){
            	   // conn.commit();
                // }
            }
            conn.commit();        
            long b=System.currentTimeMillis();  
            print("MySql依次插入"+count+"条记录到"+tableName,a,b,count);   
        } catch (Exception ex) {  
            ex.printStackTrace();  
        }finally{  
            close(conn);    
        }  
    }  
    
    /** 
     * mysql批量插入记录,point为batchsize 
     */  
    public static void insertBatch(List<tableStruct> values, String tableName,String mode,int point){  
        String sql="insert "+mode+" into "+ tableName+ " values(?,?,?,?,?)";
        Connection conn=getConn("mysql");  
        clear(conn,tableName);
        try {        
            PreparedStatement prest = conn.prepareStatement(sql);        
            long a=System.currentTimeMillis();  
            for(int i=0;i<count;i++){
                prest.setString(1,values.get(i).name);
                prest.setString(2,values.get(i).birth);        
                prest.setInt(3,values.get(i).id);
                prest.setInt(4,values.get(i).weight);
                prest.setInt(5,values.get(i).height);        
                prest.addBatch();
                if((i+1)%point==0){
                    long a1=System.currentTimeMillis();
                    prest.executeBatch();
                    long b1=System.currentTimeMillis();
                    double c=point*1000.0/(b1-a1);
                    System.out.println("执行第"+(i+1)/point+"批，每秒执行"+c+"条");
                }    
            } 
            // prest.executeBatch();      
            conn.commit();       
            long b=System.currentTimeMillis();  
            print("MySql批量插入"+count+"条记录到"+tableName,a,b,count);
        } catch (Exception ex) {  
            ex.printStackTrace();  
        }finally{  
            close(conn);    
        }  
    }

    /** 
     * mysql批量插入记录,div为划分的个数，每次重新建立连接
     */  
    public static void insertBatchReConn(String inputPath, String tableName,String mode,int div)throws IOException{
        Connection conn=getConn("mysql");  
        clear(conn,tableName);
        close(conn);
        int batchsize=count/div;

        FileReader reader = new FileReader(inputPath);
        BufferedReader br = new BufferedReader(reader);  

        long a=System.currentTimeMillis();
        for(int i=0;i<div;i++){
            List<tableStruct> values=new ArrayList<tableStruct>();
            for (int j=0;j<batchsize;j++){
                tableStruct value=new tableStruct();
                String[] str=new String[5];
                str=br.readLine().split("\\,");
                value.name=str[0].substring(1,9);
                value.birth=str[1].substring(1,11);
                value.id=Integer.parseInt(str[2]);
                value.height=Integer.parseInt(str[3]);
                value.weight=Integer.parseInt(str[4]);      
                values.add(value);
            }
            System.out.println("执行第"+(i+1)+"批");
            insertOneBatch(values,tableName,mode,batchsize);
        }
        long b=System.currentTimeMillis(); 
        print("MySql分"+div+"次批量插入"+count+"条记录到"+tableName,a,b,count);
    }

    /** 
     * 批量提交一次
     */ 
    public static void insertOneBatch(List<tableStruct> values, String tableName,String mode,int batchsize){  
        String sql="insert "+mode+" into "+ tableName+ " values(?,?,?,?,?)";
        Connection conn=getConn("mysql");  
        try {        
            PreparedStatement prest = conn.prepareStatement(sql);        
            long a=System.currentTimeMillis();  
            for(int i=0;i<batchsize;i++){
                prest.setString(1,values.get(i).name);
                prest.setString(2,values.get(i).birth);        
                prest.setInt(3,values.get(i).id);
                prest.setInt(4,values.get(i).weight);
                prest.setInt(5,values.get(i).height);        
                prest.addBatch();
            }
            prest.executeBatch();      
            conn.commit();       
            long b=System.currentTimeMillis();  
            print("MySql批量插入"+batchsize+"条记录到"+tableName,a,b,batchsize);
        } catch (Exception ex) {  
            ex.printStackTrace();  
        }finally{  
            close(conn);    
        }  
    }
    /** 
     * load文件块
     */ 
    public static void loadAll(String tableName,String mode){
        String loadSql="load data infile '/mnt/mysqlTest/data.txt' "+mode+ " into table "+tableName+ 
         " FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED BY '\"' LINES TERMINATED BY '\n'";
        Connection conn=getConn("mysql");  
        clear(conn,tableName);
        try{
            PreparedStatement prest = conn.prepareStatement(loadSql);        
            long a=System.currentTimeMillis(); 
            prest.execute();
            conn.commit();       
            long b=System.currentTimeMillis();  
            print("MySql一次load ("+mode+")"+count+"条记录到"+tableName,a,b,count);
        } catch (Exception ex) {  
            ex.printStackTrace();  
        }finally{  
            close(conn);              
        }
    }  
 
}  

File: python/mysqlTest.py

# -*- coding: utf-8 -*-

'''
可选：
set global net_buffer_length=1000000; 
set global max_allowed_packet=1000000000;

execute result:

num=4000000  table1
insertPerline
354.47 sec
insertBatch
1, 51.76 sec
5, 51.57 sec
10, 51.67 sec
20, 51.52 sec
30, 52.14 sec
50, 53.82 sec
100, 51.66 sec
200, 50.93 sec
--------------
loadAll test1   test2  test3
        11.81   326.3  locks exceed the lock table size
'''

import time
import MySQLdb


def printResults(key,ts,te):
    speed=100000/(te-ts)
    print(key+",用时 %f"%(te-ts)+" s,平均每秒执行 %d"%speed+"条")
    print("----------------------------------")

def getConn():
    try:
        ts=time.time()
        conn=MySQLdb.connect(host='localhost',user='root',passwd='cloudera',db='test',port=3306)
        te=time.time()
        print("创建连接用时%f ms"%((te-ts)*1000))
        return conn
    except MySQLdb.Error,e:
        print "Mysql Error %d: %s" % (e.args[0], e.args[1])        


def clear_table(tableName):
    try:
        conn=MySQLdb.connect(host='localhost',user='root',passwd='cloudera',db='test',port=3306)
        cur=conn.cursor()
        cur.execute("truncate %s"%tableName)
        conn.commit()
        cur.close()
        conn.close()
        print("执行清理操作成功")
    except MySQLdb.Error,e:
        print "Mysql Error %d: %s" % (e.args[0], e.args[1])

# 每次insert一句
def insertPerline(tableName,values,mode='ignore'):
    num=len(values)
    conn=getConn()
    try:
        ts=time.time()
        cur=conn.cursor()

        for i in range(num):
            stmt1="insert %s into %s"%(mode,tableName)
            stmt=stmt1+" values('%s','%s',%d,%d,%d)"%values[i]
            cur.execute(stmt)
        conn.commit()
        te=time.time()
        printResults("MySql依次插入10万条记录到"+tableName,ts,te)
    except MySQLdb.Error,e:
        print "Mysql Error %d: %s" % (e.args[0], e.args[1])
    cur.close()
    conn.close()

# 分批insert,优化
def insertBatch(tableName,div,values,mode='ignore'):
    num=len(values)
    conn=getConn()
    try:
        ts=time.time()
        cur=conn.cursor()
        # mysqldb的占位符统一写为%s,区别于python
        stmt1="insert %s into %s"%(mode,tableName) 
        stmt=stmt1+" values(%s,%s,%s,%s,%s)"
        for i in range(div):
            # 该语句对insert做专门的优化,相当于执行一条insert: 
            cur.executemany(stmt,values[num/div*i:num/div*(i+1)])
        conn.commit()
        te=time.time()
        printResults("MySql批量插入10万条记录到"+tableName,ts,te)
    except MySQLdb.Error,e:
        print "Mysql Error %d: %s" % (e.args[0], e.args[1])
        conn.rollback()
    cur.close()
    conn.close()

# 一次性load
def loadAll(tableName,mode='ignore',disable=False):
    conn=getConn()
    try:
        ts=time.time()
        cur=conn.cursor()
        stmt=("load data infile '/mnt/mysqlTest/data.txt' %s into table %s" 
         " FIELDS TERMINATED BY ',' OPTIONALLY ENCLOSED BY '\"' LINES TERMINATED BY '\n'"%(mode,tableName)
        )            
        cur.execute(stmt)
        conn.commit()
        te=time.time()
        printResults("MySql一次load ("+mode+") 10万条记录到 "+tableName,ts,te)
    except MySQLdb.Error,e:
        print "Mysql Error %d: %s" % (e.args[0], e.args[1])
        conn.rollback()
    cur.close()
    conn.close()


if __name__ == '__main__':
    tableName=['test1','test2','test3']
    
    file=open('../data.txt')
    lines=file.readlines()
    file.close()
    values=[]

    for line in lines:
        l=line.rstrip('\n').split(',')
        ll=(l[0][1:-1],l[1][1:-1],int(l[2]),int(l[3]),int(l[4]))
        values.append(ll)
    num=len(values)
    # print(num)
    # ------insert和load到有数据的表3-----
    clear_table(tableName[2])
    insertBatch(tableName[2],10,values)
    print('begin')
    insertBatch(tableName[2],10,values)
    clear_table(tableName[2])
    insertBatch(tableName[2],10,values)
    print('begin')
    loadAll(tableName[2])


    # for i in range(3):
    #     clear_table(tableName[i])
    #     insertPerline(tableName[i],values)

    #     clear_table(tableName[i])
    #     insertBatch(tableName[i],5,values)

    #     clear_table(tableName[i])
    #     loadAll(tableName[i])

    #     clear_table(tableName[i])
    #     loadAll(tableName[i],'replace')
    # insertBatch(tableName[0],5,values)
    # insertBatch(5,values,tableName)
    # loadAll(tableName[2],'ignore')
    # clear_table()
    # func_mysqldb2(values)
    # for i in [5,10,20,30,50,100,200]:
    #     clear_table()
    #     func_mysqldb2(num,i,values)
    # func_mysqldb1(num,1,values)

